mega codegen restructuring

vtjnash opened the pull request:
this rewrite of codegen is intended to help make it much more obvious to new users (and old) of how type information is flowing through the codegen system.

the key changes are that raw llvm `Value*` objects are almost never allowed (although they can be created with `boxed()` or `emit_unbox()`) and that all Intrinsics are now fully typed. in the longer term, it should be feasible to do a lot of new optimizations through this data structure :). also, when llvm removes `PointerType->getElementType()` in the next release, this should also provide a straightforward path to handling the resulting codegen upgrade necessity.

i know this is huge, but I hope this will provide a basis for adding more advanced & flexible codegen features in the future with much lower impact. I also feel that this has been needed for a long time, as the codegen system has been organically growing with support for so many new unboxed type representations.

TODO before merging:
- [x] cleanup commit history (i don't think i want to flatten it entirely into one commit, but the breakdown right now is pretty pointless)
- [x] get signoff from @carnaval
- [x] get signoff from @Keno
- [x] get signoff from @JeffBezanson 
- [x] check / cleanup any MCJIT and LLVM37 issues

TODO after merging:
- continue to reduce codegen dependence on expr_type
- remove distinction between emit_boxed / emit_unboxed
- add more argument type validation to Intrinsics so that invalid usage results in emitting an explicit error rather than potentially triggering an llvm assertion
- track alignment information more explicitly
- track allocation (gcroot) information in the jl_cgval_t representation
- write docs for codegen – this is probably a good time to sit down and write a solid description of this system for other devs
- ~~document that many Intrinsics eval their arguments at compile-time~~ remove all Intrinsics

vtjnash committed:
[wip] codegen rewrite + rewrite intrinsics to accomodate

this starts the work to replace naked Value* pointers in codegen with
information-rich jl_cgval_t objects. the intrinsics were also refactored
to propagate type information so that unboxed values would not be
necessary for that case (and all values could be first-class).

vtjnash committed:
[wip codegen rewrite] cleanup closure env handling to occur early

with this rewrite, environment argument handling can now occur in the
emit_function setup rather than dealing with it in emit_var

vtjnash committed:
[wip codegen rewrite] fix marking of varinfo.value when type_is_ghost and recursive constant struct creation

programs should never write to the .typ field directly, since it may
miss that the type is actually ghost, or should be moved to the stack

the recursive case of constant struct to llvm Value* was broken by the
latest changes to move structs to the stack. this fixes that and adds a
test case for it.

vtjnash committed:
[wip codegen rewrite] fix isVolatile marking and closure memloc allocation marking

vtjnash committed:
[wip codegen rewrite] fix select_value validation and missing value copy in variable assignment

vtjnash committed:
[wip codegen rewrite] fix mixed-type intrinsic usage (such as `bswap(Float64)`)

vtjnash committed:
[wip codegen rewrite] fix another type information mismatch error in assignment

vtjnash committed:
[wip codegen rewrite] cleanup/fix llvm_type_rewrite [ci skip]

vtjnash committed:
[wip codegen rewrite] cleanup/fix generic_box

vtjnash committed:
[wip codegen rewrite] fix elty reference in julia_to_native

vtjnash committed:
[wip codegen rewrite] rewrite jl_f_is codegen handling to add some missing runtime gcroots and handle some DataType values better

vtjnash committed:
[wip codegen rewrite] restore volatile load behavior validity

vtjnash committed:
[wip codegen rewrite] fix jl_alloca result type and alignment

vtjnash committed:
remove assertion that emitting an argument didn't result in an error (undef). this can happen

vtjnash committed:
bitstypes don't need to be leaftypes to have a size and llvm representation, so they are also allowed for generic_box

JeffBezanson commented:
This is indeed an excellent design improvement, a long time coming. Unfortunately the timing is not good at all. For example, after the tuple change --- which changed codegen a lot, but much less than this --- all tests passed and so on, but we were fixing the long tail of bugs for a couple months. I think this has to wait until after release-0.4 is branched.

ScottPJones commented:
Ditto to @JeffBezanson's comment, looks great, but safer not in release-0.4.
The only thing that *might* change my mind, is if this is at all necessary to get pre-compilation.
Screaming hordes of Julians will be at your door if they don't get pre-compilation soon :grinning:

carnaval commented:
If this does not go in 0.4 then we'll have to hold off on any non-trivial codegen thing until then to avoid making this a rebase nightmare. i.e. more reasons to branch of 0.4 soon.

Anyway, love this change, it takes us closer to many of my favorite things (such as removing .typ). I'll take time to review in depth. Good job Jameson !

ScottPJones commented:
@carnaval Good point. So, should this be massively reviewed, and put into 0.4?

tkelman commented:
The question is whether we want a sort-of-buggy 0.4 soonish, that then becomes impossible to backport anything that touches codegen, or a more buggy but easier to backport fixes to 0.4 further out.

ihnorton commented:
Do people think it is realistic to get 0.4 **out** in under, say, 1 month? If it's going to be another 2+ months then let's just damn the torpedoes. We're not 1.0 for a reason.

timholy commented:
My personal preference would be to hit merge on package precompilation and immediately tag 0.4-beta. But there are multiple paths forward.

ScottPJones commented:
I guess I'd agree with @timholy

vtjnash commented:
This is a bug fix, precompilation is a new feature. The intention behind this is simply to make codegen easier to maintain. While a large number of lines changed, the translation is nearly always one-to-one, as demonstrated by the final +/- line count. Precompilation OTOH, will have a long tail of julia, package, and user issues that will need to be addressed over the next few months before it can be declared reasonably stable.

ScottPJones commented:
If this really is a bug fix, then yes, get it in.

quinnj commented:
I believe the general rule is that bug fixes can generally *wait* since additional 0.4 releases can happen with bug fixes. New features/APIs/breaking changes, however, are important to get in before an official 0.4 branch.

ScottPJones commented:
That's the opposite of my experience, which was bug fixes should always be allowed in, and backported if possible, while new features/APIs etc. (breaking changes pretty much were not allowed) would not go end near a official release branch.  We'd have 3: release, next release, latest. When next release is considered to not regress bugwise from release, it becomes release, latest becomes next release, etc.

timholy commented:
> Precompilation OTOH, will have a long tail of julia, package, and user issues that will need to be addressed over the next few months before it can be declared reasonably stable.

The interesting thing about precompilation is that once merged into some version of julia, everyone will switch to using that version. So we merge it now and try to get the bugs ironed out ASAP, or we merge it at the beginning of 0.5 and watch everyone switch to using master throughout the entire development cycle of 0.5.

These remarks are, of course, somewhat orthogonal to the fate of this particular PR, except for the fact that two destabilizing transitions at once would be rough.

vtjnash committed:
fix undef type handling in ccall lowering

vtjnash committed:
fix emit_new_struct handling of bitstype

Jutho commented:
With summer holidays starting at various places, especially in academia which probably represents a significant part of the julia community, I guess at least some part of the Julia community might not be extremely concerned about whether 0.4 is released tomorrow, one month from now or in two months. But I could certainly be mistaken ... And the Julia developers might want to take some time off as well?

ScottPJones commented:
Time off?  What's that?  Besides, working on Julia (or using Julia) poolside is *fun*!
On the other hand, if a lot of the current julia users are taking time off, it might be the best time to get some of these things in and any regressions dealt with *before* they all come back in September, and want to get their real work done.

KristofferC commented:
@timholy makes a good point.The benefits of static compilations are so big that when it is merged, the gains of switching to the development branch will likely outweigh the disadvantages of it being a bit unstable. This means that even if we work hard to make the 0.4 release very stable it will be of little benefit since most people will want to be on dev anyway.

On the other hand, if static compilation goes in 0.4, we could say that it is currently a bit unstable but keep backporting fixes to it. 

Personally, package load time is the **only** feature I see Julia being inferior to any other programming language I have tried.

kshyatt commented:
>The benefits of static compilations are so big that when it is merged, the gains of switching to the development branch will likely outweigh the disadvantages of it being a bit unstable. This means that even if we work hard to make the 0.4 release very stable it will be of little benefit since most people will want to be on dev anyway.

I feel like this has happened to some extent with the big GC improvements, even though everyone keeps saying that you shouldn't expect packages to work well/at all on 0.4

yuyichao commented:
@vtjnash 

On llvm 3.6
```
In file included from intrinsics.cpp:52:0,
                 from codegen.cpp:1491:
ccall.cpp: 在函数‘llvm::Value* llvm_type_rewrite(llvm::Value*, llvm::Type*, llvm::Type*, bool, bool, bool, jl_codectx_t*)’中:
ccall.cpp:284:20: 错误：‘class llvm::Type’ has no member named ‘isPrimitiveType’
     if (from_type->isPrimitiveType() && target_type->getPrimitiveSizeInBits() == from_type->getPrimitiveSizeInBits()) {
```

vtjnash committed:
fix ccall attribute assignment / vector overrun bug for varargs

this was the cause of the latest appveyor failure on this PR,
although the bug itself is not related to this PR

StefanKarpinski commented:
I echo @JeffBezanson's remarks about the timing of this being quite unfortunate. We need to get 0.4 out the door and this change seems certain to cause a number of nasty, hard-to-find bugs so it should not be merged in the release stabilization window – i.e. now. @vtjnash, you're saying that this is a bug fix – what bug does it fix? I'm not sure that a refactoring qualifies, no matter how much of an improvement.

JeffBezanson commented:
Yes, calling this a bug fix is nonsense. I agree there may indeed be bugs that this fixes, but the patch doesn't look, feel, smell, or act like a bug fix. It's a major refactoring.

The argument "you can't add features late, but can fix bugs any time" is normally correct, but is misapplied here. A release doesn't exist for its own sake --- the only point of doing a release at all is to ship certain features. Once those features are in, we enter a release candidate period, so that we are not actually in the position of shipping features right after they are added.

tkelman commented:
> Once those features are in, we enter a release candidate period

Which we have not yet done for 0.4-pre.

JeffBezanson commented:
> Which we have not yet done for 0.4-pre.

Point being what, that we should work on this patch and delay 0.4-pre even further?

quinnj commented:
Here's how I imagine things:
* Get precompilation done and branch 0.4 quickly
* This could then go in with 0.5 array-mageddon
* If it turns out to merge cleanly without a lot of hassle, we consider backporting to 0.4
* We're also planning on doing a very short 0.5, so if we get to 0.6 quickly enough, we probably don't even need to worry about backporting

ViralBShah commented:
+1 to @quinnj 

I would prefer to get 0.4 quickly branched, and focus on a quick and short 0.5 release, where this can go in.

tkelman commented:
If we want the set of bugs that exist on the release-0.4 branch to be similar to the set of bugs that exist on master early on, then that's an option. There are enough bugs on master right now that still need work to track down before we'll have anything release-worthy - branching early and having too much attention focused on a nicer-to-work-on master is going to make the release harder to maintain.

A short 0.5 cycle seems highly unlikely, considering #9336 will likely be necessary to resolve by then.

JeffBezanson commented:
Another +1 to @quinnj

ScottPJones commented:
>A release doesn't exist for its own sake --- the only point of doing a release at all is to ship certain features. 
That's only one way of doing releases.  Another way is to have a constant stream of releases, and any features that are ready are put in the "upcoming release" branch, which then only has bug fixes (and not major refactoring bug fixes) until it becomes the current release branch.

ScottPJones commented:
@quinnj's plan does seem reasonable, whatever the length of the 0.5 cycle.

vtjnash committed:
remove an unused gc root when vaStack is true

vtjnash committed:
type-inference propagates type information incorrectly when inlining getfield of a tuple, try to work around that in codegen. also, disable optimization of maintaining boxes, until the gc-roots for that optimization are being created correctly

vtjnash committed:
intrinsics should use staticeval_bitstype not julia_type_to_llvm since even non-leaf types (like Ptr) should be converted to integers

carnaval committed:
build with llvm trunk

carnaval commented on src/ccall.cpp:
```diff
@@ -280,33 +275,34 @@ Value *llvm_type_rewrite(Value *v, Type *from_type, Type *target_type, bool toju
+    }
+
+    // simple integer and float widening & conversion cases
+    if (from_type->isPrimitiveType() && target_type->getPrimitiveSizeInBits() == from_type->getPrimitiveSizeInBits()) {
```
isPrimitiveType has apparently disappeared from LLVM since late 2013

carnaval commented on src/cgutils.cpp:
```diff
@@ -1486,20 +1338,20 @@ static Value *emit_getfield_knownidx(Value *strct, unsigned idx, jl_datatype_t *
         Value *addr = builder.CreateConstInBoundsGEP2_32(
-            cast<PointerType>(strct->getType()->getScalarType())->getElementType(),
+            cast<PointerType>(strct.V->getType()->getScalarType())->getElementType(),
             strct, 0, idx);
```
missing .V here

vtjnash replied:
you're welcome to push fixes directly to this branch; i've pushed all of my local work.

carnaval replied:
k. only small fixes to get it to build on llvm trunk. it could probably use a rebase but i'll let you do that later ;-)

ArchRobison commented:
Nice work!  I have questions about `jl_cgval_t` in  `codegen.cpp`.  Since Github won't show me the diff (too large), I'll quote the relevant section below:
```
    bool isboxed; // whether this value is a jl_value_t* allocated on the heap with the right type tag
    bool isghost; // whether this value is "ghost"
    bool ispointer; // whether this value is actually pointer to the value
    bool isimmutable; // V points to something that is definitely immutable (e.g. not stack allocated)
```
For the comment about `isboxed`, should it really say `jl_value_t allocated on the heap`?  I.e., it's not a pointer being allocated on the heap, is it?  Does `isboxed` always imply `ispointer`?

What is a ghost?  The comments for `jl_cgval_t` seems like a relevant place to explain the concept.

The phrase `definitely immutable` could use clarification as to the lifetime of the immutability.  Does it mean something that does not change until it becomes unreachable by GC?  Or a stronger condition than that?

vtjnash commented:
some of those are essentially just placeholders for future work.

> For the comment about isboxed, should it really say jl_value_t allocated on the heap? I.e., it's not a pointer being allocated on the heap, is it? Does isboxed always imply ispointer?

yes, isboxed always implies ispointer (I didn't allow multiple pointer indirections). there might be a missing comma; the full sentence could also be written as: `// whether ".V" is a jl_pvalue_llvmt pointer to the object, which has been allocated on the heap with the right type tag`

> What is a ghost? The comments for jl_cgval_t seems like a relevant place to explain the concept.

a singleton immutable object, such as "nothing". it doesn't show up in the llvm code, so it's given the name "ghost"

> The phrase definitely immutable could use clarification as to the lifetime of the immutability. Does it mean something that does not change until it becomes unreachable by GC? Or a stronger condition than that?

the lifetime of a jl_cgval_t is bounded to the function being emitted

yuyichao commented on src/intrinsics.cpp:
```diff
@@ -652,12 +638,14 @@ static Value *emit_pointerref(jl_value_t *e, jl_value_t *i, jl_codectx_t *ctx)
+            return jl_cgval_t();
         }
         assert(jl_is_datatype(ety));
         uint64_t size = jl_datatype_size(ety);
```
@vtjnash  Should the `CreateCall` of `jlallocobj_func` be replaced by `emit_allocobj` as well?

vtjnash committed:
fix emit_unbox of bool to only read the defined bit

vtjnash committed:
assigning to the cgval.typ field is invalid

all modifications of cgval need to go through the constructor to
maintain the expected invariants

vtjnash committed:
fix missing optimization and gc root condition test for emit_f_is (fixes dates test)

vtjnash committed:
fix static_constant_instance to handle bool correctly

vtjnash committed:
[wip codegen rewrite] use correct macro test for jl_is_immutable

vtjnash committed:
not all who are constants are static

vtjnash committed:
UnixDomainCM connect timeout is stricter than necessary, also it is best to wait on the output not the process to avoid forcing the kernel to buffer the data

vtjnash committed:
avoid creating gc roots for isghost variables

vtjnash committed:
in debug builds, give the parts of the gcframe nice names for easier LLVM IR readability

vtjnash commented:
when someone gets a chance, the first 4 commits and the relevant parts of https://github.com/JuliaLang/julia/commit/8da287267faa738f41c49fc2de20df8a3f4839b6 should be moved to master

Keno commented:
I have a number of fixes on top of this which I will push soon. However, I am about to go to sleep, if somebody could look at the following test case:

```
for j = 1:1
    x = try
        error()
        2
    catch
        continue
    end
end
```
The problem is that there is a GenSym that is read but never assigned to in a dead branch:
```
AST(:($(Expr(:lambda, Any[], Any[Any[Any[symbol("#s1"),:Any,2],Any[:j,:Any,18],Any[:x,:Any,2]],Any[],4,Any[]], :(begin  # none, line 2:
        GenSym(0) = (Main.colon)(1,1)
        #s1 = (top(start))(GenSym(0))
        unless (top(!))((top(done))(GenSym(0),#s1)) goto 1
        2:
        GenSym(1) = (top(next))(GenSym(0),#s1)
        j = (top(getfield))(GenSym(1),1)
        #s1 = (top(getfield))(GenSym(1),2) # none, line 3:
        $(Expr(:enter, 4)) # none, line 4:
        (Main.error)() # none, line 5:
        GenSym(2) = 2
        x = GenSym(2)
        $(Expr(:leave, 1))
        goto 5
        4:
        $(Expr(:leave, 1))
        goto 3
        x = GenSym(3)
        5:
        3:
        unless (top(!))((top(!))((top(done))(GenSym(0),#s1))) goto 2
        1:
        0:
        return
    end)))))
```
So maybe the correct solution here is to get rid of that code at a higher level (@JeffBezanson?).

Keno committed:
Fix build failures and errors in the test suite

Keno committed:
Comment out an assertion and add a test case that triggers it

Keno committed:
Re-export julia_type_to_llvm

Keno commented:
I put up a preliminary branch with some changes at kf/codegen_rewrite. For now I just commented out the assertion that gets triggered by the case above, but it would be good if it can be fixed properly.

vtjnash commented:
removing the assertion sounds reasonable to me. or you might consider a read from a gensym to also set the assigned flag (assigned to undef), to continue trying to catch any out-of-order usages.

vtjnash commented:
since @Keno and @carnaval (and myself) have done other work based against this branch, I think that counts as testing. Anyone plan on doing a line-by-line review, or should I rebase one last time and merge this?

StefanKarpinski commented:
I think merging into master is the only way to get this tested out.

ArchRobison commented:
+1 for merging.

Keno commented:
Do include the few changes I had on that branch above, other than that +1 to merging from me as well.

ArchRobison commented:
I'm not sure what key I bumped on my keyboard 30 seconds ago, but I certainly did not intend to close the issue when I commented.  I had intended to say I looked over the patch a while back and it really did look like a one-for-one restructuring with low chance of causing problems.

tkelman commented:
We branched early and there's a lot of other 0.5 work held up on this, so +1 from me. Looks unlikely to have many conflicts with other changes that will need backporting to release-0.4.

vtjnash commented:
that sounds unanimous, so i'll merge in the morning when CI finishes (@keno, i've included your changes too. thanks for reminding me to grab those).

tkelman commented:
Exciting segfault while running `doc/genstdlib.jl`: http://buildbot.e.ip.saba.us:8010/builders/build_osx10.9-x64/builds/850/steps/shell_1/logs/stdio

vtjnash commented:
Indeed. Type inference is generating some strange code for
```julia
Base.getindex(Base.ObjectIdDict, IntrinsicFunction)
```
```julia
julia> code_typed(Base.getindex, (Base.ObjectIdDict, IntrinsicFunction))
1-element Array{Any,1}:
 :($(Expr(:lambda, Any[:t,:key], Any[Any[Any[:t,ObjectIdDict,0],Any[:key,IntrinsicFunction,0],Any[:v,Any,18]],Any[],Any[],Any[]], :(begin  # dict.jl, line 282:
        v = (top(ccall))(:jl_eqtable_get,Base.Any,(top(svec))(Base.Any,Base.Any,Base.Any)::SimpleVector,(top(getfield))(t::ObjectIdDict,:ht)::Array{Any,1},0,key,0,Base.secret_table_token,0) # dict.jl, line 283:
        unless (Base.is)(v,Base.secret_table_token)::Bool goto 0 # dict.jl, line 284:
        (Base.throw)($(Expr(:new, :((top(getfield))(Base,:KeyError)::Type{KeyError}), :(key::_))))::Union{}
        0:  # dict.jl, line 286:
        return v
    end))))
```
```julia
281 function getindex(t::Associative, key)
282     v = get(t, key, secret_table_token)
283     if is(v, secret_table_token)
284         throw(KeyError(key))
285     end
286     return v
287 end
```
for some reason, inference has not marked the right type for `key` and this is throwing an assertion failure in the new codegen (due to not expecting inference to put a `TypeVar` as the `Expr.typ`)

tkelman commented:
> patched on master via 3bf729141e88657bbc39ebcd99161337d9e504e5

Cool. If we are going to aim to backport this for 0.4.1 or 0.4.2, it would be extremely valuable to keep a list somewhere of all follow-up bug fix commits.

vtjnash commented:
good point. any thoughts on the best way to do that? i'm expecting to try to cause the .cpp files to rapidly diverge further, so `git log *.cpp` may not be the best plan

tkelman commented:
> so `git log *.cpp` may not be the best plan

No, probably not, history gets murky quickly if you're just working from logs. Would prefer we try to keep backporting-relevant notes in a consistent place, written down as the commits are made on master. Probably a new issue with an itemized list? Same applies to any makefile fixes we might need for the out-of-tree builds.

vtjnash commented:
maybe it's worth turning back on the github wiki for this sort of tracking stuff?

mbauman commented:
Push the bugfix commits onto the `jn/codegen_rewrite` branch as a temporary staging area?  Or better yet, open a new branch based on 0.4 and maybe even the backport PR right now?

tkelman commented:
I'm worried that merge fingers will get prematurely itchy, but I am much more used to tracking WIP via PR's than github wiki's.

JeffBezanson commented:
Good to merge this into master, but I don't yet see a compelling reason to backport it. If we encounter a bug in 0.4 *in practice* that's fixed by this, we can weigh at that point whether it's better to backport or do a more minimal patch against 0.4.

StefanKarpinski commented:
At some point it will make backporting bug fixes easier if this is backported. Otherwise patches won't apply, etc. But sure, we can wait until that's an actual problem before we try to backport this change.

ScottPJones commented:
What about bringing Elvis to Julia?

JeffBezanson commented:
BTW, this has a large number of WIP commits and could have used some squashing.

vtjnash commented:
The WIP tag is to identify that those commits were incomplete (in case someone tries to bisect this too much), but were left distinct because each was addressing a different part of the transition (so that git blame can be a little bit more informative)
